lib/workflow/postgres/postgres.go

```diff
@@ -0,0 +1,726 @@
+	event := &backend.WorkflowEvent{}
+	workc := &PostgresWorkflowCompleter{
+		ctx:             ctx,
+		enqueueTaskStmt: stmt,
+		tx:              tx,
+		cancel:          cancel,
+	}
+	var taskResultID sql.NullInt64
+	err = row.Scan(
+		&workc.wid,
+		&event.Instance.InstanceName,
+		&event.Instance.WorkflowName,
+		&event.Instance.Status,
+		&event.Instance.Parameters,
+		&event.Instance.Payload,
+		&workc.eid,
+		&event.Type,
+		&taskResultID,
+		&event.EnqueuedTaskCount,
+		&event.CompletedTaskCount,
+	)
+	if err == sql.ErrNoRows {
+		tx.Commit() // nolint: errcheck
+		cancel()
+		return nil, nil, workflow.ErrNoWorkflowInstances
+
+	} else if err != nil {
+		cancel()
+		return nil, nil, errors.Wrap(err, "failed to dequeue workflow")
+	}
+
+	event.InstanceID = workc.wid
+
+	if event.Type == backend.TaskComplete {
+		event.CompletedTaskCount++
+		row := tx.QueryRowContext(ctx, getTaskResultQuery, taskResultID)
+		tr := backend.TaskResult{}
+		err = row.Scan(
+			&tr.TaskName, &tr.Parameters, &tr.Status, &tr.ErrorText, &tr.Result)
+		if err != nil {
+			cancel()
+			// FIXME FIXME FIXME
+			// TODO is this an infinite loop (this is panic situation)
+			return nil, nil, err
+		}
+		event.TaskResult = &tr
+	}
+
+	workc.enqueuedTaskCount = event.EnqueuedTaskCount
+	workc.completedTaskCount = event.CompletedTaskCount
+
+	return event, workc, nil
+}
+
+func (workc *PostgresWorkflowCompleter) EnqueueTask(task *backend.Task, opts backend.TaskEnqueueOpts) error {
+	if opts.TryRemaining <= 0 {
+		opts.TryRemaining = 1
+	}
+	if opts.StartAfter.IsZero() {
+		opts.StartAfter = time.Now()
+	}
+
+	_, err := workc.enqueueTaskStmt.ExecContext(workc.ctx,
+		task.WorkflowInstanceID, opts.TryRemaining, opts.StartAfter, task.Name, task.Parameters)
+	if err != nil {
+		return errors.Wrap(err, "failed to enqueue task")
+	}
+
+	workc.enqueuedTaskCount = workc.enqueuedTaskCount + 1
+	return nil
+}
+
+func (pg *PostgresBackend) dequeueTask(tx *sql.Tx, taskName string) (int64, *backend.Task, error) {
+	row := tx.QueryRow(dequeueTaskQuery, taskName)
+	task := &backend.Task{
+		Name: taskName,
+	}
+
+	var tid int64
+	err := row.Scan(&tid, &task.WorkflowInstanceID, &task.Parameters)
+	if err == sql.ErrNoRows {
+		return 0, nil, workflow.ErrNoTasks
+	} else if err != nil {
+		return 0, nil, errors.Wrap(err, "failed to dequeue task")
+	}
+	return tid, task, nil
+}
+func (pg *PostgresBackend) DequeueTask(ctx context.Context, taskName string) (*backend.Task, backend.TaskCompleter, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	tx, err := pg.db.BeginTx(ctx, nil)
+	if err != nil {
+		cancel()
+		return nil, nil, errors.Wrap(err, "failed to dequeue task")
+	}
+
+	taskc := &PostgresTaskCompleter{
+		ctx:    ctx,
+		_tx:    tx,
+		cancel: cancel,
+	}
+
+	tid, task, err := pg.dequeueTask(tx, taskName)
+	if err != nil {
+		cancel()
+		return nil, nil, err
+	}
+	taskc.tid = tid
+
+	// pinger
+	go func() {
```

Comment:
Let's move this into a named function so it is easier to find when we inevitably are reading through dlv output.